syntax = "proto3";

package fedlearner.common;

message RegressionTreeNodeProto {
    int32 node_id = 1;
    int32 left_child = 2;
    int32 right_child = 3;
    int32 parent = 4;

    bool is_owner = 5;
    int32 owner_id = 6;
    int32 feature_id = 7;
    bool is_cat_feature = 11;
    float threshold = 8;
    repeated int32 cat_threshold = 12;
    bool default_left = 10;

    float weight = 9;
}

message PartitionInfo {
    int32 num_partitions = 1;
};

message RegressionTreeProto {
    repeated RegressionTreeNodeProto nodes = 1;
    repeated float feature_importance = 3;
};

message BoostingTreeEnsambleProto {
    repeated RegressionTreeProto trees = 1;
    repeated float feature_importance = 2;
    repeated string feature_names = 3;
    repeated string cat_feature_names = 4;
};

message EncryptedNumbers {
    repeated bytes ciphertext = 1;
}

message Histograms {
    repeated EncryptedNumbers hists = 2;
}

message SplitInfo {
    int32 node_id = 1;
    float gain = 2;
    int32 owner_id = 3;
    int32 feature_id = 4;
    repeated int32 split_point = 5;
    bool default_left = 10;
    float left_weight = 6;
    float right_weight = 7;
    repeated int32 left_samples = 8;
    repeated int32 right_samples = 9;
}

message VerifyParams {
    repeated string example_ids = 1;
    float learning_rate = 2;
    int32 max_iters = 3;
    int32 max_depth = 4;
    int32 max_leaves = 5;
    float l2_regularization = 6;
    int32 max_bins = 7;
    string grow_policy = 8;
    bool validation = 9;
    int32 num_trees = 10;
    bool leader_no_data = 11;
}

// For  hists
message values_normal{
  bytes values = 1;
  repeated uint32 shape = 2;
  string dtype = 3;
}

message values_paillier{
  repeated bytes values = 1;
  repeated uint32 shape=2;
}

message histogram_request_leader{
  uint32 base = 1;
  values_normal values = 2;
  values_normal binned_features = 3;
  repeated uint32 thresholds = 4;
}

message histogram_request_follower{
  uint32 base = 1;
  values_paillier values = 2;
  values_normal binned_features = 3;
  repeated uint32 thresholds = 4;
  bytes public_key = 5;
}

message histogram_leader{
  repeated values_normal hists =1;
}
message histogram_follower{
  repeated values_paillier hists =1;
}

message decrypt_histogram_request{
  uint32 base = 1;
  repeated bytes keys = 3;
  repeated EncryptedNumbers hists = 4;
}
message decrypt_histogram_reply{
  repeated values_normal hists=1;
}

// For encrypt
message  encrypt_numbers_request{
  repeated float numbers = 1;
  bytes public_key = 2;
}


service HistsService {
  rpc ComputeHists_leader (histogram_request_leader) returns (histogram_leader) {}
  rpc ComputeHists_follower (histogram_request_follower) returns (histogram_follower) {}
  rpc Decrypt_histogram (decrypt_histogram_request) returns (decrypt_histogram_reply) {}
  rpc Encrypt_numbers (encrypt_numbers_request) returns (EncryptedNumbers) {}
}